__tests__/fetch_request.js

```diff
@@ -0,0 +1,234 @@
+    })
+    test('is not set by formData', () => {  
+      const formData = new FormData()
+      formData.append("this", "value")
+      const formDataRequest = new FetchRequest("get", "localhost", { body: formData })
+      expect(formDataRequest.fetchOptions.headers)
+        .toStrictEqual(defaultHeaders)
+    })
+    test('is set by file body', () => {      
+      const file = new File(["contenxt"], "file.txt", { type: "text/plain" })
+      const fileRequest = new FetchRequest("get", "localhost", { body: file })
+      expect(fileRequest.fetchOptions.headers)
+        .toStrictEqual({ ...defaultHeaders, "Content-Type": "text/plain"})    
+    })
+    test('is set by json body', () => {      
+      const jsonRequest = new FetchRequest("get", "localhost", { body: { some: "json"} })
+      expect(jsonRequest.fetchOptions.headers)
+        .toStrictEqual({ ...defaultHeaders, "Content-Type": "application/json"})
+    })
+  })
+
+  test('additional headers are appended', () => {
+    const request = new FetchRequest("get", "localhost", { contentType: "application/json", headers: { custom: "Header" } })
+    expect(request.fetchOptions.headers)
+      .toStrictEqual({ ...defaultHeaders, custom: "Header", "Content-Type": "application/json"})
+    request.addHeader("test", "header")
+    expect(request.fetchOptions.headers)
+      .toStrictEqual({ ...defaultHeaders, custom: "Header", "Content-Type": "application/json", "test": "header"})
+  })    
+
+  test('headers win over contentType', () => {
+    const request = new FetchRequest("get", "localhost", { contentType: "application/json", headers: { "Content-Type": "this/overwrites" } })
+    expect(request.fetchOptions.headers)
+      .toStrictEqual({ ...defaultHeaders, "Content-Type": "this/overwrites"})
+  })    
+
+  test('serializes JSON to String', () => {
+    const jsonBody = { some: "json" }
+    let request
+    request = new FetchRequest("get", "localhost", { body: jsonBody, contentType: "application/json" })
+    expect(request.fetchOptions.body).toBe(JSON.stringify(jsonBody))
+
+    request = new FetchRequest("get", "localhost", { body: jsonBody })
+    expect(request.fetchOptions.body).toBe(JSON.stringify(jsonBody))
+  })
+
+  test('not serializes JSON with explicit different contentTyp', () => {
+    const jsonBody = { some: "json" }
+    const request = new FetchRequest("get", "localhost", { body: jsonBody, contentType: "not/json" })
+    expect(request.fetchOptions.body).toBe(jsonBody)
+  })
+
+  test('set redirect', () => {
+    let request
+    const redirectTypes = [ "follow", "error", "manual" ]
+    for (const redirect of redirectTypes) {
+      request = new FetchRequest("get", "localhost", { redirect })
+      expect(request.fetchOptions.redirect).toBe(redirect)
+    }
+    
+    request = new FetchRequest("get", "localhost")
+    expect(request.fetchOptions.redirect).toBe("follow")
+
+    // maybe check for valid values and default to follow?
```

Comment:
Isn't the default being checked in the prior check? About checking valid values, I don't think that's necessary, since Fetch API will raise an error if the user provides an invalid value. At least I think that we would be testing Fetch API behavior and not Request.JS behavior. Having a test that proves that it's possible to override the default value provided by request.js is enough, like the test bellow (even if that isn't a valid value)